### PR TITLE
Mlse hotfix

### PR DESCRIPTION
**Pull request by kingsotn-twelve**
p0: 

- [x]  preflight checks #90 
- [x] interrupt function on terminal #88 
    - [x] Add to event state management
- [x] Create and test a dedicated error structure
- [x] implement error structure, then handle the errors in top level function calls  #91 
    - [x] implement for all tool calls
    ~~- [ ] implement for all nodes~~ did not want to overengineer here
- [x] human in the loop #87 for tool calling. example is before combine_clips correclty identify clips we want to combine

### CODE REVIEW

**Comment by kingsotn-twelve** [comment]
there are no api calls to check how many credits you have. the only way to see usage is when the completion request returns a `.usage` field in the json, but that isn't helpful for checking credits

so i created a simple preflight check with [error codes via openai](https://platform.openai.com/docs/guides/error-codes/api-errors)

**Comment by kingsotn-twelve** [comment]
MLSE wanted to not propagate every error back up to supervisor. If say ffmpeg fails (or any tool/execution workers) then break out of more executions immediately instead of allowing supervisor to hallucinate a fix.

This does not mean sysexit though, we are to carefully return to terminal chatting and maintain the State(). how the frontend handles this failure is arbitrary — either read the ERROR_MSG and fail, or parse what the "next_steps" are.

1. create the TOOL_ERROR states (see langgraph errors too) use pydanticv1 namespace
2. exit gracefully in video_editing.py back to Chat:
3. maintain the State() where it will contain some kind of Error(..fail_reason: TOOL_FAILURE, tool: VIDEO_EDITING)

I will first implement for all @tool decorated functions

via [jockey/stirrups/video_editing.py](https://github.com/twelvelabs-io/tl-jockey/blob/1e61bc5ceda39d41c3eb81f6e172b200027db738/jockey/stirrups/video_editing.py)

**Comment by kingsotn-twelve** [comment]
As I've continued to work on this i've created a [JockeyError](https://github.com/twelvelabs-io/tl-jockey/blob/mlse-hotfix/jockey/stirrups/errors.py) for jockey specific errors. Langgraph has their own errors [(docs)](https://langchain-ai.github.io/langgraph/reference/errors/) [(code)](https://github.com/langchain-ai/langgraph/blob/main/libs/langgraph/langgraph/errors.py) that I have implemented in relevant functions. By the end of this deep dive i should have fixed (or at least have addressed) many of the issues in our repo.

### Todos
- [x] create a custom JockeyError in errors.py to handle Jockey specific errors
- [x] init pytest
- [x] video_editing.py error handling and testing in tests/
    - [x] test_combine_clips_success
    - [x] test_combine_clips_download_error 
- [x] stirrup.py implement basic langgraph errors [here](https://github.com/twelvelabs-io/tl-jockey/blob/mlse-hotfix/jockey/stirrups/stirrup.py)

- [ ] wrap all TypeErrors with JockeyError so that `Exceptions` are properly handled in the call stack [(example)](https://github.com/twelvelabs-io/tl-jockey/blob/1e61bc5ceda39d41c3eb81f6e172b200027db738/jockey/jockey_graph.py#L91)
- [ ] fix recursion problems and repetitiveness of all nodes
- [ ] handle the langgraph errors in stirrup.py such as what to do when graph recursion limit is reached
- [ ] error handling in:
    - [ ] _worker_node()
    - [ ] _planner_node()
    - [ ] _reflect_node()
- [ ] pytest test_stirrup.py (non-priority)

**Comment by kingsotn-twelve** [comment]
Interrupt events are now pushed to the events in `astream_events()`. The schema for `on_error` events look like the following. The JockeyError and LanggraphError structures are similar. [See those functions here.](https://github.com/twelvelabs-io/tl-jockey/pull/93/commits/28c2efb25ab376e5a0065d9e381f01abbb638559#diff-b328375db70d5c506d8c6d9bae8621cc0fbfaf2c49b9a7aadc277c4cf820d816)

```json
# schema example
{
    "event": "on_interrupt",
    "name": "JockeyInterrupt",
    "run_id": "string",
    "data": {
        "message": "Stream interrupted by user",
        "last_event": {
            "event": "string",
            "metadata": {
                "langgraph_node": "string"
            }
        },
        "event_type": "string",
        "node": "string"
    },
    "tags": ["string"],
    "metadata": {
        "interrupted_at": {
            "langgraph_node": "string"
        }
    }
}

```

and here's an example of what it looks like populated with actual data

```json
# schema with data example
{
    "event": "on_interrupt",
    "name": "JockeyInterrupt",
    "run_id": "e665dcf4-2dc4-4d55-b60b-9a0faebe2f1f",
    "data": {
        "message": "Stream interrupted by user",
        "last_event": {
            "event": "on_chat_model_stream",
            "data": {
                "chunk": {
                    "content": "",
                    "id": "run-3c0903dd-6d25-4910-bee7-87cba6c57a8a"
                }
            },
            "run_id": "3c0903dd-6d25-4910-bee7-87cba6c57a8a",
            "name": "AzureChatOpenAI",
            "tags": [
                "seq:step:2",
                "planner"
            ],
            "metadata": {
                "langgraph_step": 4,
                "langgraph_node": "planner",
                "langgraph_triggers": [
                    "branch:supervisor:condition:planner"
                ],
                "langgraph_path": [
                    "__pregel_pull",
                    "planner"
                ],
                "langgraph_checkpoint_ns": "planner:1141019b-43ae-6736-dbbd-23aee7be026a",
                "checkpoint_ns": "planner:1141019b-43ae-6736-dbbd-23aee7be026a",
                "ls_provider": "azure",
                "ls_model_name": "gpt-4o",
                "ls_model_type": "chat",
                "ls_temperature": 0.0
            },
            "parent_ids": [
                "3684681d-4f34-4785-b162-3d100314de5d",
                "ed797c44-3b51-495a-ae21-7340ff302e4e",
                "ec0b7065-08b2-4aff-b531-0186f324464f"
            ]
        },
        "event_type": "on_chat_model_stream",
        "node": "planner"
    },
    "tags": [
        "seq:step:2",
        "planner"
    ],
    "metadata": {
        "interrupted_at": {
            "langgraph_step": 4,
            "langgraph_node": "planner",
            "langgraph_triggers": [
                "branch:supervisor:condition:planner"
            ],
            "langgraph_path": [
                "__pregel_pull",
                "planner"
            ],
            "langgraph_checkpoint_ns": "planner:1141019b-43ae-6736-dbbd-23aee7be026a",
            "checkpoint_ns": "planner:1141019b-43ae-6736-dbbd-23aee7be026a",
            "ls_provider": "azure",
            "ls_model_name": "gpt-4o",
            "ls_model_type": "chat",
            "ls_temperature": 0.0
        }
    }
}

```

**Comment by kingsotn-twelve** [comment]
The following diagram demonstrates how `JockeyError` propagates through the system, with `combine_clips`, `Stirrup._call_tools`, and `Worker Layer` grouped as part of the **Worker** section.

We can print errors to CLI via try except handling, but because errors are also sent through the **event_stream** we can also parse those events in our frontend. An example of how we do it is to collect the stream to our frontend via [streamEventsApis.js](https://github.com/twelvelabs-io/tl-jockey/blob/1e61bc5ceda39d41c3eb81f6e172b200027db738/frontend/src/apis/streamEventsApis.js#L4)

```mermaid
sequenceDiagram
    box gray Worker
        participant ToolFunction as Tool Function (combine_clips)
        participant Stirrup as Stirrup._call_tools (base_tool.ainvoke)
        participant Worker as Worker<br>worker_prompt | worker_llm_with_tools | self._call_tools
    end
    participant EventStream as Event Stream<br>jockey.astream_events()

    ToolFunction ->> Stirrup: Raises JockeyError
    Stirrup ->> Worker: Catches and re-raises error
    Worker ->> EventStream: Propagates error
    EventStream ->> EventStream: Catches in try/except block
    EventStream ->> CLI: Prints error to CLI

File: jockey/stirrups/video_search.py [hunk 1]
```diff
@@ -135,38 +143,38 @@ async def simple_video_search(
...
+                    node=NodeType.WORKER,
+                    error_type=ErrorType.SEARCH,
+                    function_name=WorkerFunction.VIDEO_SEARCH,
+                    details=f"Error: {str(error)}",
+                )
+                jockey_error_event = create_jockey_error_event(error=jockey_error)
+                await parse_langchain_events_terminal(jockey_error_event)
+                raise jockey_error
         else:
             print(f"Search error: {search_results}")
```
Comment (ojadeyemi) [inline jockey/stirrups/video_search.py:159]: this line seems to print search result even when the search was successful.

**Comment by kingsotn-twelve** [comment]
langsmith is helpful for debugging and easy to setup. also shows nice-to-have token and cost usage, and time to first token (ttfs) which is a super important metric in most ai apps today. highly recommend.

Screenshot 2024-10-31 at 11 19 30 PM

**Comment by kingsotn-twelve** [comment]
> langsmith is helpful for debugging and easy to setup. also shows nice-to-have token and cost usage, and time to first token (ttfs) which is a super important metric in [most ai apps today ](https://www.youtube.com/watch?v=e-gwvmhyU7A&t=8041s)(although we don't control that). highly recommend.
> 
> Screenshot 2024-10-31 at 11 19 30 PM

How to setup langsmith?

1. signup/login on smith.langchain.com
2. go to home and click Set up tracing button
3. With langchain -> generate api key
4. copy this into your .env and replace `key` and `project name` with what was given
```yaml
LANGCHAIN_TRACING_V2=true
LANGCHAIN_ENDPOINT="https://api.smith.langchain.com"
LANGCHAIN_API_KEY="<your-api-key>"
LANGCHAIN_PROJECT="pr-upbeat-ceramics-52"
```
5. when you run `python3 -m jockey` you should be able to see your trace

**Comment by kingsotn-twelve** [comment]
for human in the loop, ryan from mlse suggested:
- dynmaic worker prompting to customize for each user
- context window replacing per step instead of appending to save context window.

my thoughts:
- not the most practical solution, can just use RAG if personalization is important. simpler solution is to clearly outline the limitations of jockey to humans.
- could work, but what if we need to continously prompt the human >> in that case it would be nice to have some history to prevent looping — context window increase may be inevitable

**Comment by kingsotn-twelve** [comment]
Screenshot 2024-11-04 at 6 51 49 PM

**Comment by kingsotn-twelve** [comment]
new and improved!
Screenshot 2024-11-06 at 6 01 14 PM

**Comment by kingsotn-twelve** [comment]
[video demo update on human-in-the-loop feature](https://www.loom.com/share/9d0a6e610698470ebf759c702f3bccb9?sid=8e60cfd9-59e9-4a44-b3dd-158eccd04bc1)
[diagram](https://excalidraw.com/#json=rFoZItB9f0dy8g2fuJ61F,OCZLOklj9IOV-LjEM-CfXQ)
tldr:
showcase architecture and flow of code
planner works with human feedback
workers break every now and then... need more time to make robust

**Comment by kingsotn-twelve** [comment]
i will start working on streaming human in the loop events for frontend support

example schema of the event:
```{
    'event': 'on_human_feedback_request',
    'data': {
        'input': {
            'question': 'Should this action proceed?',
            'context': {...},
            'options': ['approve', 'reject', 'modify']
        }
    },
    'name': 'HumanFeedback',
    'tags': ['human_loop'],
    'run_id': '...',
    'metadata': {...},
    'parent_ids': [...]
}
```

**Comment by kingsotn-twelve** [comment]
1. problem was prompt templates weren't [correctly escaping](https://python.langchain.com/docs/troubleshooting/errors/INVALID_PROMPT_INPUT/), bc they are passed as type(str). [here is the fix](https://github.com/twelvelabs-io/tl-jockey/pull/93/commits/c563d5af18519d8befaf9ebfb5338c3c9753f580)
2.  problem was [`as_node` usage](https://langchain-ai.github.io/langgraph/reference/graphs/#langgraph.graph.graph.CompiledGraph.aget_state_history) was wrong. routing is now fixed completely for the ask_human node including reflect node. the basic idea is that as_node="ask_human" means that it skips the ask_human(), and returns the ask_human() return value as if it came from that node. [here is the fix](https://github.com/twelvelabs-io/tl-jockey/pull/93/commits/7f7f4859d79dd51cc16cb2769200c88b1ef5db4d)

Here are some example images that prove routing now works very robustly. and [here is the code for that](https://github.com/twelvelabs-io/tl-jockey/pull/93/commits/3a193483975f57c0f57882a6569c5ae77bbc2eb9)

- planner -> video-search
  Screenshot 2024-11-26 at 5 44 57 PM

- video-search -> video-editing
  Screenshot 2024-11-26 at 5 45 19 PM

however, these spawned other problems im working on:
- [ ] states aren't being passed correctly between instructor and worker tool calls
- [ ] video-editing worker is broken (was mostly always broken), but that's gna change after i start working on it.

**Comment by kingsotn-twelve** [comment]
refactors and improvements for cleaner graph state. the above commits also squashes some bugs related to video-editing tool worker

Screenshot 2024-11-27 at 6 39 31 PM
